extension-ui/background.js

```diff
@@ -1,84 +1,139 @@
-
 let blockedSitesCache = null;
+let allowedSitesCache = null;
+let isBlackList = true;
 
-// Initialize cache when the extension is loaded
-chrome.runtime.onStartup.addListener(() => initializeBlockedSitesCache());
-chrome.runtime.onInstalled.addListener(() => initializeBlockedSitesCache());
+chrome.runtime.onStartup.addListener(() => initializeCaches());
+chrome.runtime.onInstalled.addListener(() => initializeCaches());
 
-function initializeBlockedSitesCache(callback) {
-  chrome.storage.local.get("blockedSites", (data) => {
+function initializeCaches(callback) {
+  chrome.storage.local.get(["blockedSites", "allowedSites", "isBlackList"], (data) => {
     blockedSitesCache = data.blockedSites || [];
+    allowedSitesCache = data.allowedSites || [];
+    isBlackList = data.isBlackList !== undefined ? data.isBlackList : true;
     if (typeof callback === "function") {
       callback();
     }
   });
 }
 
-// Use this function to ensure cache is initialized
-function ensureBlockedSitesCacheInitialized(callback) {
-  if (blockedSitesCache === null) {
-    initializeBlockedSitesCache(callback);
+function ensureCachesInitialized(callback) {
+  if (blockedSitesCache === null || allowedSitesCache === null) {
+    initializeCaches(callback);
   } else if (typeof callback === "function") {
     callback();
   }
 }
 
 chrome.webNavigation.onBeforeNavigate.addListener((details) => {
-  ensureBlockedSitesCacheInitialized(() => {
+  ensureCachesInitialized(() => {
     handleBeforeNavigate(details);
   });
 }, { url: [{ schemes: ['http', 'https'] }] });
 
 function handleBeforeNavigate(details) {
   try {
     const url = new URL(details.url);
-    
     if (url.protocol === 'chrome:' || url.protocol === 'about:') {
       return;
     }
 
     const hostname = url.hostname.toLowerCase();
-    
-    if (blockedSitesCache.some(site => hostname.includes(site))) {
-      chrome.tabs.get(details.tabId, (tab) => {
-        if (tab.url.startsWith('chrome://') || tab.url.startsWith('about:')) {
-          return;
-        }
-        chrome.scripting.executeScript({
-          target: { tabId: details.tabId },
-          func: () => {
-            //TODO  add UI for the oops window
-             window.stop();
-            window.location.href = chrome.runtime.getURL('oops.html');
-          }
-        }).catch(error => {
-          console.error("Error executing script: ", error);
-        });
-      });
+
+    if (isBlackList) {
+      if (blockedSitesCache.some(site => hostname.includes(site))) {
+        blockSite(details.tabId);
+      }
+    } else {
+      if (allowedSitesCache.some(site => hostname.includes(site))) {
```

Comment:
check only  if (!allowedSitesCache.some(site => hostname.includes(site))) without the return;